File: src/main/java/com/mewna/catnip/entity/message/Embed.java
/*
 * Copyright (c) 2018 amy, All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holder nor the names of its contributors
 *    may be used to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.mewna.catnip.entity.message;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.mewna.catnip.entity.impl.EmbedImpl;
import com.mewna.catnip.entity.impl.EmbedImpl.*;
import lombok.Getter;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.time.OffsetDateTime;
import java.util.List;

/**
 * All methods of this class (and inner classes) have nullable returns, unless otherwise stated.
 *
 * @author natanbc
 * @since 9/2/18.
 */
@SuppressWarnings("unused")
@JsonDeserialize(as = EmbedImpl.class)
public interface Embed {
    /**
     * Embed title.
     *
     * @return Title of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    String title();
    
    /**
     * Embed type.
     *
     * @return Type of the embed. Never null.
     */
    @Nonnull
    @CheckReturnValue
    EmbedType type();
    
    /**
     * Embed description.
     *
     * @return Description of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    String description();
    
    /**
     * Embed URL.
     *
     * @return URL of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    String url();
    
    /**
     * Embed timestamp.
     *
     * @return Timestamp of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    OffsetDateTime timestamp();
    
    /**
     * Embed color.
     *
     * @return Color of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Integer color();
    
    /**
     * Embed footer.
     *
     * @return Footer of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Footer footer();
    
    /**
     * Embed image.
     *
     * @return Image of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Image image();
    
    /**
     * Embed thumbnail.
     *
     * @return Thumbnail of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Thumbnail thumbnail();
    
    /**
     * Embed video.
     *
     * @return Video of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Video video();
    
    /**
     * Embed provider.
     *
     * @return Provider of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Provider provider();
    
    /**
     * Embed author.
     *
     * @return Author of the embed, or null if absent.
     */
    @Nullable
    @CheckReturnValue
    Author author();
    
    /**
     * Embed fields.
     *
     * @return Fields of the embed. Never null.
     */
    @Nonnull
    @CheckReturnValue
    List<Field> fields();
    
    enum EmbedType {
        IMAGE("image"),
        VIDEO("video"),
        LINK("link"),
        RICH("rich"),
        UNKNOWN("");
        
        @Getter
        private final String key;
        
        EmbedType(final String key) {
            this.key = key;
        }
        
        @Nonnull
        @CheckReturnValue
        public static EmbedType byKey(@Nonnull final String key) {
            for(final EmbedType type : values()) {
                if(type.key.equals(key)) {
                    return type;
                }
            }
            return UNKNOWN;
        }
    }
    
    @JsonDeserialize(as = AuthorImpl.class)
    interface Author {
        /**
         * Name of the author.
         *
         * @return Embed author's name. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String name();
        
        /**
         * URL of the author.
         *
         * @return Embed author's URL, or null if absent.
         */
        @CheckReturnValue
        String url();
        
        /**
         * Icon URL of the author.
         *
         * @return Embed author's icon URL, or null if absent.
         */
        @CheckReturnValue
        String iconUrl();
        
        /**
         * Icon proxy URL of the author.
         *
         * @return Embed author's icon proxy URL, or null if absent.
         */
        @CheckReturnValue
        String proxyIconUrl();
    }
    
    @JsonDeserialize(as = FieldImpl.class)
    interface Field {
        /**
         * Name of the field.
         *
         * @return Field's name. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String name();
        
        /**
         * Value of the field.
         *
         * @return Field's value. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String value();
        
        /**
         * Whether the field is inline.
         *
         * @return True if the field is inline, false otherwise.
         */
        @CheckReturnValue
        boolean inline();
    }
    
    @JsonDeserialize(as = FooterImpl.class)
    interface Footer {
        /**
         * Text of the footer.
         *
         * @return Footer's text. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String text();
        
        /**
         * Icon URL of the footer.
         *
         * @return Footer's icon URL, or null if absent.
         */
        @CheckReturnValue
        String iconUrl();
        
        /**
         * Icon proxy URL of the footer.
         *
         * @return Footer's icon proxy URL, or null if absent.
         */
        @CheckReturnValue
        String proxyIconUrl();
    }
    
    @JsonDeserialize(as = ImageImpl.class)
    interface Image {
        /**
         * URL of the image.
         *
         * @return Image's URL. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String url();
        
        /**
         * Proxy URL of the image.
         *
         * @return Image's proxy URL, or null if absent.
         */
        @CheckReturnValue
        String proxyUrl();
        
        /**
         * Height of the image.
         *
         * @return Image's height, or -1 if absent.
         */
        @CheckReturnValue
        int height();
        
        /**
         * Width of the image.
         *
         * @return Image's width, or -1 if absent.
         */
        @CheckReturnValue
        int width();
    }
    
    @JsonDeserialize(as = ProviderImpl.class)
    interface Provider {
        /**
         * Name of the provider.
         *
         * @return Provider's name, or null if absent.
         */
        @CheckReturnValue
        String name();
        
        /**
         * URL of the provider.
         *
         * @return Provider's URL. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String url();
    }
    
    @JsonDeserialize(as = ThumbnailImpl.class)
    interface Thumbnail {
        /**
         * URL of the thumbnail.
         *
         * @return Thumbnail's URL. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String url();
        
        /**
         * Proxy URL of the thumbnail.
         *
         * @return Thumbnail's proxy URL, or null if absent.
         */
        @CheckReturnValue
        String proxyUrl();
        
        /**
         * Height of the thumbnail.
         *
         * @return Thumbnail's height, or -1 if absent.
         */
        @CheckReturnValue
        int height();
        
        /**
         * Width of the thumbnail.
         *
         * @return Thumbnail's width, or -1 if absent.
         */
        @CheckReturnValue
        int width();
    }
    
    @JsonDeserialize(as = VideoImpl.class)
    interface Video {
        /**
         * URL of the video.
         *
         * @return Video's URL. Never null.
         */
        @Nonnull
        @CheckReturnValue
        String url();
        
        /**
         * Height of the video.
         *
         * @return Video's height, or -1 if absent.
         */
        @CheckReturnValue
        int height();
        
        /**
         * Width of the video.
         *
         * @return Video's width, or -1 if absent.
         */
        @CheckReturnValue
        int width();
    }
}

File: packages/discord.js/src/structures/Embed.js
'use strict';

const { embedLength } = require('@discordjs/builders');
const isEqual = require('fast-deep-equal');

/**
 * Represents an embed.
 */
class Embed {
  constructor(data) {
    /**
     * The API embed data.
     *
     * @type {APIEmbed}
     * @readonly
     */
    this.data = { ...data };
  }

  /**
   * An array of fields of this embed.
   *
   * @type {Array<APIEmbedField>}
   * @readonly
   */
  get fields() {
    return this.data.fields ?? [];
  }

  /**
   * The title of this embed.
   *
   * @type {?string}
   * @readonly
   */
  get title() {
    return this.data.title ?? null;
  }

  /**
   * The description of this embed.
   *
   * @type {?string}
   * @readonly
   */
  get description() {
    return this.data.description ?? null;
  }

  /**
   * The URL of this embed.
   *
   * @type {?string}
   * @readonly
   */
  get url() {
    return this.data.url ?? null;
  }

  /**
   * The color of this embed.
   *
   * @type {?number}
   * @readonly
   */
  get color() {
    return this.data.color ?? null;
  }

  /**
   * The timestamp of this embed. This is in an ISO 8601 format.
   *
   * @type {?string}
   * @readonly
   */
  get timestamp() {
    return this.data.timestamp ?? null;
  }

  /**
   * @typedef {Object} EmbedAssetData
   * @property {?string} url The URL of the image
   * @property {?string} proxyURL The proxy URL of the image
   * @property {?number} height The height of the image
   * @property {?number} width The width of the image
   */

  /**
   * The thumbnail of this embed.
   *
   * @type {?EmbedAssetData}
   * @readonly
   */
  get thumbnail() {
    if (!this.data.thumbnail) return null;
    return {
      url: this.data.thumbnail.url,
      proxyURL: this.data.thumbnail.proxy_url,
      height: this.data.thumbnail.height,
      width: this.data.thumbnail.width,
    };
  }

  /**
   * The image of this embed.
   *
   * @type {?EmbedAssetData}
   * @readonly
   */
  get image() {
    if (!this.data.image) return null;
    return {
      url: this.data.image.url,
      proxyURL: this.data.image.proxy_url,
      height: this.data.image.height,
      width: this.data.image.width,
    };
  }

  /**
   * The video of this embed.
   *
   * @type {?EmbedAssetData}
   * @readonly
   */
  get video() {
    if (!this.data.video) return null;
    return {
      url: this.data.video.url,
      proxyURL: this.data.video.proxy_url,
      height: this.data.video.height,
      width: this.data.video.width,
    };
  }

  /**
   * @typedef {Object} EmbedAuthorData
   * @property {string} name The name of the author
   * @property {?string} url The URL of the author
   * @property {?string} iconURL The icon URL of the author
   * @property {?string} proxyIconURL The proxy icon URL of the author
   */

  /**
   * The author of this embed.
   *
   * @type {?EmbedAuthorData}
   * @readonly
   */
  get author() {
    if (!this.data.author) return null;
    return {
      name: this.data.author.name,
      url: this.data.author.url,
      iconURL: this.data.author.icon_url,
      proxyIconURL: this.data.author.proxy_icon_url,
    };
  }

  /**
   * The provider of this embed.
   *
   * @type {?APIEmbedProvider}
   * @readonly
   */
  get provider() {
    return this.data.provider ?? null;
  }

  /**
   * @typedef {Object} EmbedFooterData
   * @property {string} text The text of the footer
   * @property {?string} iconURL The URL of the icon
   * @property {?string} proxyIconURL The proxy URL of the icon
   */

  /**
   * The footer of this embed.
   *
   * @type {?EmbedFooterData}
   * @readonly
   */
  get footer() {
    if (!this.data.footer) return null;
    return {
      text: this.data.footer.text,
      iconURL: this.data.footer.icon_url,
      proxyIconURL: this.data.footer.proxy_icon_url,
    };
  }

  /**
   * The accumulated length for the embed title, description, fields, footer text, and author name.
   *
   * @type {number}
   * @readonly
   */
  get length() {
    return embedLength(this.data);
  }

  /**
   * The hex color of this embed.
   *
   * @type {?string}
   * @readonly
   */
  get hexColor() {
    return typeof this.data.color === 'number'
      ? `#${this.data.color.toString(16).padStart(6, '0')}`
      : (this.data.color ?? null);
  }

  /**
   * Returns the API-compatible JSON for this embed.
   *
   * @returns {APIEmbed}
   */
  toJSON() {
    return { ...this.data };
  }

  /**
   * Whether the given embeds are equal.
   *
   * @param {Embed|APIEmbed} other The embed to compare against
   * @returns {boolean}
   */
  equals(other) {
    if (other instanceof Embed) {
      return isEqual(this.data, other.data);
    }

    return (
      this.author?.iconURL === other.author?.icon_url &&
      this.author?.name === other.author?.name &&
      this.author?.url === other.author?.url &&
      this.color === (other.color ?? null) &&
      this.description === (other.description ?? null) &&
      this.footer?.iconURL === other.footer?.icon_url &&
      this.footer?.text === other.footer?.text &&
      this.image?.url === other.image?.url &&
      this.thumbnail?.url === other.thumbnail?.url &&
      (this.timestamp && Date.parse(this.timestamp)) === (other.timestamp ? Date.parse(other.timestamp) : null) &&
      this.title === (other.title ?? null) &&
      this.url === (other.url ?? null) &&
      this.video?.url === other.video?.url &&
      isEqual(this.fields, other.fields?.map(field => ({ ...field, inline: field.inline ?? false })) ?? []) &&
      isEqual(this.provider, other.provider ?? null)
    );
  }
}

exports.Embed = Embed;
